GivTCP/read.py

```diff
@@ -446,147 +444,194 @@ def getData(fullrefresh):      #Read from Invertor put in cache
+
+        multi_output = calcBatteryValue(multi_output)
 
-        multi_output=calcBatteryValue(multi_output)
-        
-        if 'multi_output_old' in locals(): 
-            multi_output=dataCleansing(multi_output,regCacheStack[4])
+        if 'multi_output_old' in locals():
+            multi_output = dataCleansing(multi_output, regCacheStack[4])
         # only update cache if its the same set of keys as previous (don't update if data missing)
 
-        if 'multi_output_old' in locals(): 
-            MOList=dicttoList(multi_output)
-            MOOList=dicttoList(multi_output_old)
-            dataDiff=set(MOOList) - set(MOList)
-            if len(dataDiff)>0:
+        if 'multi_output_old' in locals():
+            MOList = dicttoList(multi_output)
+            MOOList = dicttoList(multi_output_old)
+            dataDiff = set(MOOList) - set(MOList)
+            if len(dataDiff) > 0:
                 for key in dataDiff:
                     logger.critical(str(key)+" is missing from new data, publishing all other data")
 
         # Add new data to the stack
         regCacheStack.pop(0)
         regCacheStack.append(multi_output)
-        # Save new data to Pickle
-        with open(GivLUT.regcache, 'wb') as outp:
-            pickle.dump(regCacheStack, outp, pickle.HIGHEST_PROTOCOL)
-        result['result']="Success retrieving data"
 
-        # Success, so delete oldDataCount
-        if exists(GivLUT.oldDataCount):
-            os.remove(GivLUT.oldDataCount)
+        # Get lastupdate from pickle if it exists
+        with cacheLock:
+            if exists(GivLUT.lastupdate):
+                with open(GivLUT.lastupdate, 'rb') as inp:
+                    previousUpdate = pickle.load(inp)
+                timediff = datetime.datetime.fromisoformat(multi_output['Last_Updated_Time'])-datetime.datetime.fromisoformat(previousUpdate)
+                multi_output['Time_Since_Last_Update'] = (((timediff.seconds*1000000)+timediff.microseconds)/1000000)
+
+            # Save new time to pickle
+            with open(GivLUT.lastupdate, 'wb') as outp:
+                pickle.dump(multi_output['Last_Updated_Time'], outp, pickle.HIGHEST_PROTOCOL)
+
+            # Save new data to Pickle
+            with open(GivLUT.regcache, 'wb') as outp:
+                pickle.dump(regCacheStack, outp, pickle.HIGHEST_PROTOCOL)
+                
+            logger.info("Successfully retrieved from: " + GiV_Settings.invertorIP)
+
+            result['result'] = "Success retrieving data"
+
+            # Success, so delete oldDataCount
+            if exists(GivLUT.oldDataCount):
+                os.remove(GivLUT.oldDataCount)
 
     except:
-        GivLUT.consecFails()
+        consecFails()
+
         e = sys.exc_info()
         logger.error("Error processing registers: " + str(e))
         logger.error("Invertor Update failed so using last known good data from cache ("+previousUpdate+")")
```

Comment:
`previousUpdate` might to be set for instance at start up or when the cache gets deleted due to consecutive read failures

observed this myself

`2023-01-30T03:59:34.17861239Z stderr F Traceback (most recent call last):
2023-01-30T03:59:34.1787344Z stderr F   File "/app/GivTCP_1/read.py", line 414, in getData
2023-01-30T03:59:34.209365959Z stderr F     battery['Battery_SOC'] = multi_output_old['Battery_Details'][b.battery_serial_number]['Battery_SOC']
2023-01-30T03:59:34.20946824Z stderr F KeyError: '\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
2023-01-30T03:59:34.209505553Z stderr F 
2023-01-30T03:59:34.209543223Z stderr F During handling of the above exception, another exception occurred:
2023-01-30T03:59:34.209577496Z stderr F 
2023-01-30T03:59:34.209621424Z stderr F Traceback (most recent call last):
2023-01-30T03:59:34.209657604Z stderr F   File "/app/GivTCP_1/read.py", line 919, in <module>
2023-01-30T03:59:34.210579093Z stderr F     globals()[sys.argv[1]]()
2023-01-30T03:59:34.210636254Z stderr F   File "/app/GivTCP_1/read.py", line 603, in self_run2
2023-01-30T03:59:34.211318131Z stderr F     runAll("False")
2023-01-30T03:59:34.211382683Z stderr F   File "/app/GivTCP_1/read.py", line 558, in runAll
2023-01-30T03:59:34.211902853Z stderr F     getData(full_refresh)
2023-01-30T03:59:34.212046679Z stderr F   File "/app/GivTCP_1/read.py", line 524, in getData
2023-01-30T03:59:34.218949675Z stderr F     logger.error("Invertor Update failed so using last known good data from cache ("+previousUpdate+")")
2023-01-30T03:59:34.21911484Z stderr F UnboundLocalError: local variable 'previousUpdate' referenced before assignment`

Reply:
Ah! Good spot. So do we just need some null handling here for `previousUpdate` do you think?

Reply:
Oh! Rather no, it's that the variable is actually undefined. Because, Python. I shall fix, cheers.

Reply:
looking at this now, so if you can update PR with this fix then I can merge.

Reply:
Currently modbus calls are managed via the REDIS queue, and I have on my list to check if lockfiles (and now your new Lock method) are even needed, as there can't be concurrent modbus calls as the redis queue only has one worker. Thoughts?

Reply:
The error comes in when logging the lack of update. There's no need to log the detail of previousUpdate, so I have removed that from the code and can merge this now.

Reply:
> Currently modbus calls are managed via the REDIS queue, and I have on my list to check if lockfiles (and now your new Lock method) are even needed, as there can't be concurrent modbus calls as the redis queue only has one worker. Thoughts?

This is true actually. I think I was thrown because I'd seen other non-enqueued calls to `getData` like from the REST get call (but admittedly that's only in a first-run, and it's also only if the data is missing.)

Is the purpose of Redis to enable single-queues even if there are multiple instances of GivTCP running? Or, is it purely there to act as a way of enforcing one concurrent Inverter operation?

Cheers

Reply:
Good point on REST not calling getData via the queue, I'll look at that.
The intent was to use the queue to avoid multiple modbus call to a single invertor across both read and write functions. Enforcing a queue on all modbus calls should minimise clashes on read/write where one is async and other is regularly polled. I'm open to the most simple resolution to this. If a shared lock object can be used across both read and write then that might be more elegant? (I do use the queue for scheduled tasks like reverting forceCharge etc... too)

Reply:
, It doesn't call `getData` from the REST get call any more, that's OK. It was a real edge-case anyway.

Yes, you're right, maybe we could think about Lock rather than Redis for all the calls? I guess the gold-standard would be to abstract this in some way, so that the caller doesn't need to know whether it's using Redis or indeed any other method?

Currently I think there are still a few race conditions in some of the `write` operations:

* Some use the modbus client directly, rather than through the Redis instance.
* Some access the `regcache` file; maybe they should be using the Lock to ensure it's finished being written to elsewhere.

Of course Redis is also used to doing scheduled operations, e.g. forced timed charge. We still need it for that. And arguably if Redis is in use, then it's not safe to rely on a critical section Lock because Redis is running in a separate process I think? (not familiar with running a Redis process in Python. Full disclosure: I'm a C# developer, not a Python one!) 

p.s. if there's appetite for more convo on this kind of thing, maybe we could move it out of the PR discussion and onto the new Discord server?

Reply:
Sure, whats the Discord Server?

Reply:
Full credit joshgreenuk 🥇   https://community.givenergy.cloud/d/1342-new-givenergy-inverter-bms-firmware-beta-testing-chat/664